src/query-builder.js

```diff
@@ -0,0 +1,186 @@
+   * @param  {...args} args     Arguments passed to query builder.
+   * @returns {QueryBuilder}    Builder class.
+   */
+  _makeQuery(boolType, ...args) {
+    let nested = _.last(args)
+
+    let newQuery = this._buildQuery(...args)
+
+    if (_.isFunction(nested)) {
+      let clause = newQuery[_.findKey(newQuery)]
+      let builder = new QueryBuilder()
+      let recursiveResult = nested(builder)
+      if (!_.isEmpty(recursiveResult._aggs)) {clause.aggs = recursiveResult._aggs}
+      if (!_.isEmpty(recursiveResult._queries)) {clause.query = recursiveResult._queries}
+      if (!_.isEmpty(recursiveResult._filters)) {clause.filter = recursiveResult._filters}
+
+      return newQuery
+    }
+
+    return boolMerge(newQuery, this._queries, boolType)
+  }
+
+  _buildQuery(type, field, value, opts) {
+    let clause = {}
+
+    if (field && value && opts) {
+      clause = _.merge({[field]: value}, opts)
+    } else if (field && value) {
+      clause = {[field]: value}
+    } else if (field) {
+      clause = {field}
+    }
+
+    return {[type]: clause}
+  }
+
+  _makeFilter(boolType, ...args) {
+    let nested = _.last(args)
+
+    let newFilter = this._buildFilter(...args)
+
+    if (_.isFunction(nested)) {
+      let clause = newFilter[_.findKey(newFilter)]
+      let builder = new QueryBuilder()
+      let recursiveResult = nested(builder)
+
+      if (!_.isEmpty(recursiveResult._aggs)) {clause.aggs = recursiveResult._aggs}
+      if (!_.isEmpty(recursiveResult._queries)) {clause.query = recursiveResult._queries}
+      if (!_.isEmpty(recursiveResult._filters)) {clause.filter = recursiveResult._filters}
+
+      return newFilter
+    }
+
+    return boolMerge(newFilter, this._filters, boolType)
+  }
+
+  _buildFilter(type, field, value, opts) {
+    let clause = {}
+
+    if (field && value && opts) {
+      clause = _.merge({[field]: value}, opts)
+    } else if (field && value) {
+      clause = {[field]: value}
+    } else if (field) {
+      clause = {field}
+    }
+
+    return {[type]: clause}
+  }
+
+  _makeAggregation(...args) {
+    let nested = _.last(args)
+
+    let newAggregation = this._buildAggregation(...args)
+
+    if (_.isFunction(nested)) {
+      let clause = newAggregation[_.findKey(newAggregation)]
+      let builder = new QueryBuilder()
+      let recursiveResult = nested(builder)
+      if (!_.isEmpty(recursiveResult._aggs)) {clause.aggs = recursiveResult._aggs}
+      if (!_.isEmpty(recursiveResult._queries)) {clause.query = recursiveResult._queries}
+      if (!_.isEmpty(recursiveResult._filters)) {clause.filter = recursiveResult._filters}
+      return newAggregation
+    }
+
+    return newAggregation
+  }
+
+  _buildAggregation(type, field, name, opts) {
+    if (_.isObject(name)) {
+      let tmp = opts
+      opts = name
+      name = tmp
+    }
+
+    name = name || `agg_${type}_${field}`
+
+    return {
+      [name]: {
+        [type]: (() => _.assign({field}, opts))()
```

Comment:
in this case `field` can be overwritten by `opts`. Would rather expect
```js
Object.assign({}, opts, {field})
```

(I also think I understood that `_.assign` is deprecated in favour of `Object.assign`, but I might be mistaken)